misk-feature/src/main/kotlin/misk/feature/MoshiExt.kt

```diff
@@ -0,0 +1,37 @@
+package misk.feature
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonDataException
+import misk.logging.getLogger
+
+private val logger = getLogger<FeatureFlags>()
+
+/**
+ * Attempts to use [JsonAdapter.failOnUnknown] and logs any issues before falling back to ignoring
+ * the unknown fields.
+ */
+fun <T> JsonAdapter<T>.toSafeJson(value: T): String {
+  return try {
+    failOnUnknown().toJson(value)
+  } catch (e: JsonDataException) {
+    logger.error(e) {
+      "failed to serialize JSON due to unknown fields. ignoring those fields and trying again"
+    }
+    return toJson(value)
+  }
+}
+
+/**
+ * Attempts to use [JsonAdapter.failOnUnknown] and logs any issues before falling back to ignoring
+ * the unknown fields.
+ */
+fun <T> JsonAdapter<T>.fromSafeJson(json: String): T? {
+  return try {
+    failOnUnknown().fromJson(json)
+  } catch (e: JsonDataException) {
+    logger.error(e) {
+      "failed to parse JSON due to unknown fields. ignoring those fields and trying again"
+    }
+    return fromJson(json)
```

Comment:
This will still fail if they neglect to specify a field and the type is `Type!`, right? Guess our hands are tied in that case (goes back to the debate of potentially confusing/unsafe default values versus exception throwing)

Reply:
no it actually doesn't, it just sets the value to `null` and most likely an NPE will happen when you to access it in KT

Reply:
so moshi is actually breaking the "null type safety" of kotlin :(

Reply:
actually that's not true if you use the `KotlinJsonAdapterFactory`. See https://github.com/cashapp/misk/pull/1354/files